File: disable_stderr.py
import os
import sys

class _DisableStderr(object):

  def __enter__(self):
    self._dev_null_fd = os.open('/dev/null', os.O_WRONLY)
    self._stderr_fd = os.dup(sys.stderr.fileno())
    os.dup2(self._dev_null_fd, sys.stderr.fileno())

  def __exit__(self, *args):
    os.dup2(self._stderr_fd, sys.stderr.fileno())
    os.close(self._stderr_fd)
    os.close(self._dev_null_fd)

sys.stderr.write('hoge\n')
with _DisableStderr():
  sys.stderr.write('fuga\n')
sys.stderr.write('hige\n')

File: disable_stderr.cc
#include <assert.h>
#include <fcntl.h>
#include <unistd.h>

#include <iostream>

struct NullStderrScope {
    NullStderrScope() {
        null_fd = open("/dev/null", O_WRONLY);
        assert(0 != null_fd);
        backup_fd = dup(STDERR_FILENO);
        assert(0 <= backup_fd);
        assert(STDERR_FILENO == dup2(null_fd, STDERR_FILENO));
    }

    ~NullStderrScope() {
        assert(STDERR_FILENO == dup2(backup_fd, STDERR_FILENO));
        assert(0 == close(backup_fd));
        assert(0 == close(null_fd));
    }

    int null_fd;
    int backup_fd;
};

int main() {
    std::cerr << "start\n";
    {
        NullStderrScope null_stderr;
        std::cerr << "test\n";
    }
    std::cerr << "end\n";
}
